### ISSUE TITLE
RAM issues with computation on relatively small graphs 

### ISSUE THREAD

**Issue by jmcs100** [deepmind/graph_nets#81]
Hi all! I've been trying to run a training set of graphs each with 400ish nodes, 7000 edges and a 1x8 global feature vector through a standard Graph Network and have been unable to even start training because the RAM requirements of the graphs seem to be massive (I've had this confirmed by the internal computing cluster team at my company). The latest errors I'm receiving are in the actual training step, and are of the form:

`W tensorflow/core/framework/op_kernel.cc:1401] OP_REQUIRES failed at cast_op.cc:109 : Resource exhausted: OOM when allocating tensor with shape[21367800,2700] and type int32 on /job:localhost/replica:0/task:0/device:CPU:0 by allocator mklcpu
`

This seems like a massive tensor and I'm wondering whether that's causing the issue. Is this unusual for a graph of this size, or is there an issue with the computers I'm using at my end? I've tried training on CPUs with 50GB memory allowance and on remote GPUs but the results are always the same.

Any other tips for getting round this would be really appreciated if convenient.

**Comment by alvarosg** [comment]
Usually the highest leading dimension corresponds to the average number of edges per graph multiplied times the batch dimension, what is the batch size you are using? Also if you could provide the name of the tensor of that size, that would be useful for debugging.

The fact that this happens with an int32 tenor makes me think that it is related to this issue: #38 and to [utils_tf.repeat](https://github.com/deepmind/graph_nets/blob/6f33ee4244ebe016b4d6296dd3eb99625fd9f3af/graph_nets/utils_tf.py#L510) method, which creates a matrix of integers, that could have that size. We are waiting for tensorflow to merge [this PR](https://github.com/tensorflow/tensorflow/pull/26517) and expose a more efficient implementation of `tf.repeat`, and then replace our version of `utils_tf.repeat` by that version. In the meantime you could try replacing `utils_tf.repeat` by in your version of the graph_nets library by:

```
from tensorflow.python.ops.ragged import ragged_util
def repeat(tensor, repeats, axis=0, name="repeat"):
  return ragged_util.repeat(tensor, repeats, axis=axis, name=name)
```

**Comment by jmcs100** [comment]
Thanks for getting back to me! Your answer has been very helpful - my batch size for that task was only 8, but after you said how that tensor size comes about, I chased up the batch size creation in the code. It looks like I was creating an issue because my create_placeholders() wasn't taking into account the batch size properly before the feed dictionary was set up during each iteration. The model is now running once I changed that around a bit. If you want I can still try to chase up the name of the tensor though!
